FCHL19 with periodic boundary conditions and revised OpenMP parallelization

kvkarandashev opened the pull request:
Adds support for periodic boundary conditions when generating FCHL19 representations. Adds the verification script for checking PBC works correctly (temporarily placed into `tests`).

`facsf.f90` revised to avoid using REDUCTION construct in FCHL19's OpenMP parallelization while minimizing the additional CPU and memory usage compared to a serial implementation. Admittedly, I have doubts about how necessary including "!$OMP CRITICAL" was, but I cannot see a better alternative.

Unfortunately, I had trouble getting `qmllib` to work with OpenMP for some reason, hence correctness of `fascf.f90`'s parallelization was verified by using the same file in [my fork of original `qml`](https://github.com/kvkarandashev/qml).

charnley commented:
> Unfortunately, I had trouble getting `qmllib` to work with OpenMP for some reason
can you expand on this? Did you have to remove the compiler flags or notice that it wasn't allocating all the resources?

kvkarandashev commented:
To compile with OpenMP I edited several lines in `_compile.py` as follows:
```
COMPILER_FLAGS = ["-O3", "-fopenmp", "-m64", "-march=native", "-fPIC",
                     "-Wno-maybe-uninitialized", "-Wno-unused-function", "-Wno-cpp"]
extra_flags = ["-lgomp", "-lpthread", "-lm", "-ldl"] + COMPILER_FLAGS

flags = ["-L/usr/lib/", "-lblas", "-llapack"] + extra_flags
```
After compiling the code I compared its performance on a test problem to performance of the `qml` fork with the same `fascf.f90` file. `qmllib` never seemed to use more than one core, as supported by the timings:
OMP_NUM_THREADS=1:
qml: 0.7533 mins
qmllib: 0.8178 mins
OMP_NUM_THREADS=20:
qml: 0.1390 mins
qmllib: 0.8145 mins

kvkarandashev commented:
Found a mistake in the way gradients for PBC were calculated and verified.
Fixed the verification script and added verification via finite difference.

charnley commented:
qmllib is finally updated to work with numpy2.0 and openmp flags are fixed. 

- Can you pull from master? 
- Use pathlib.Path instead of os and glob
- You are calling `fgenerate_fchl_acsf` with an extra argument `natoms_tot`, but it doesn't seem to be any changes in the fortran subroutine

kvkarandashev commented:
I successfully pulled from main and replaced references to `glob` with `pathlib.Path`; everything seems to work now.

charnley commented on tests/verification_fchl19_pbc.py:
Can you rename the test file to start with `test_`? It looks like the file was skipped during the test run.

kvkarandashev commented:
I renamed the script and also changed it in order its CPU time was more in line with the other test scripts. TBH I am still not happy about the test’s CPU time consumption, but I do not see a way to make it better without augmenting `conftest.py` with functions from its [qml2 analogue](https://github.com/qml2code/qml2-dev/blob/main/tests/conftest.py).
